### PR TITLE
tests: Create or use existing properly initialized NodeContexts

### PR DESCRIPTION
**Pull request by dongcarl**
This is part 1/n of the effort to [de-globalize `ChainstateManager`](https://github.com/bitcoin/bitcoin/pull/20158)

Reviewers: Looking for tested/Code-Review/plain-ACKs

### Context

In many of our tests, we manually instantiate `NodeContext`s or `ChainstateManager`s in the test code, which is error prone. Instead, we should create or use existing references because:
1. Before we [de-globalize `ChainstateManager`](https://github.com/bitcoin/bitcoin/pull/20158), much of our code still acts on `g_chainman` (our global `ChainstateManager`), sometimes even when you're calling a method on a specific instance of `ChainstateManager`! This means that we may act on two instances of `ChainstateManager`, which is most likely not what we want.
2. Using existing references (initialized by the `{Basic,}TestingSetup` constructors) means that you're acting on objects which are properly initialized, instead of "just initialized enough for this dang test to pass". Also, they're already there! It's free!
3. By acting on the right object, we also allow the review-only assertions in future commits of [de-globalize `ChainstateManager`](https://github.com/bitcoin/bitcoin/pull/20158) to work and demonstrate correctness.

Some more detailed debugging notes can be found in the first commit, reproduced below:
```
Previously, the validation_chainstatemanager_tests test suite
instantiated its own duplicate ChainstateManager on which tests were
performed.

This wasn't a problem for the specific actions performed in
that suite. However, the existence of this duplicate ChainstateManager
and the fact that many of our validation static functions reach for
g_chainman, ::Chain(state|)Active means we may end up acting on two
different CChainStates should we write more extensive tests in the
future.

This change adds a new ChainTestingSetup which performs all
initialization previously done by TestingSetup except:

1. Mempool sanity check frequency setting
2. ChainState initialization
3. Genesis Activation
4. {Ban,Conn,Peer}Man initialization

Means that we will no longer need to initialize a duplicate
ChainstateManger in order to test the initialization codepaths of
CChainState and ChainstateManager.

Lastly, this change has the additional benefit of allowing for
review-only assertions meant to show correctness to work in future work
de-globalizing g_chainman.

In the test chainstatemanager_rebalance_caches, an additional
LoadGenesisBlock call is added as MaybeReblanaceCaches eventually calls
FlushBlockFile, which tries to access vinfoBlockFile[nLastBlockFile],
which is out of bounds when LoadGenesisBlock hasn't been called yet.

-----

Note for the future:

In a previous version of this change, I put ChainTestingSetup between
BasicTestingSetup and TestingSetup such that TestingSetup inherited from
ChainTestingSetup.

This was suboptimal, and showed how the class con/destructor inheritance
structure we have for these TestingSetup classes is probably not the
most suitable abstraction. In particular, for both TestingSetup and
ChainTestingSetup, we need to stop the scheduler first before anything
else. Otherwise classes depending on the scheduler may be referenced
by the scheduler after said classes are freed. This means that there's
no clear parallel between our teardown code and C++'s destructuring
order for class hierarchies.

Future work should strive to coalesce (as much as possible) test and
non-test init codepaths and perhaps structure it in a more fail-proof
way.
```

### CODE REVIEW

**Comment by jnewbery** [comment]
Concept ACK. Thanks for moving ChainTestingSetup into validation_chainstatemanager_tests.cpp!

**Comment by practicalswift** [comment]
Concept ACK

**Review by ryanofsky — APPROVED** [review]
Code review ACK 39387634834a1ef06e587b30fa42a781ce45e460. Obvious ACK. This is just test changes and the first 4 commits from #20158 with commit descriptions tweaked and improved

File: src/test/validation_chainstatemanager_tests.cpp
Comment (jnewbery) [inline src/test/validation_chainstatemanager_tests.cpp:24]: You don't need these ctor arguments. Just remove them and call `BasicTestingSetup()` (which will use the default arguments for its constructor).

File: src/test/validation_chainstatemanager_tests.cpp
Comment (jnewbery) [inline src/test/validation_chainstatemanager_tests.cpp:29]: Not needed. Remove.

File: src/test/validation_chainstatemanager_tests.cpp
Comment (jnewbery) [inline src/test/validation_chainstatemanager_tests.cpp:31]: I'd suggest moving this below the "We have to run..." comment, since it fits with the code there.

File: src/test/validation_chainstatemanager_tests.cpp
Comment (jnewbery) [inline src/test/validation_chainstatemanager_tests.cpp:57]: Remove this line.

File: src/test/validation_chainstatemanager_tests.cpp
Comment (jnewbery) [inline src/test/validation_chainstatemanager_tests.cpp:9]: Remove

**Review by jnewbery — COMMENTED** [review]
utACK 39387634834a1ef06e587b30fa42a781ce45e460

Some suggestions to remove unneeded code inline.

File: src/test/validation_chainstatemanager_tests.cpp
Comment (dongcarl) [reply src/test/validation_chainstatemanager_tests.cpp:57]: I copied this from `~TestingSetup`, is there a reason why this is needed there and not here?

Comment (MarcoFalke) [reply src/test/validation_chainstatemanager_tests.cpp:57]: It's a bit odd to have the low level code copied here, and require modifying `src/test/validation_chainstatemanager_tests.cpp` whenever the testing setup is upgraded.

wouldn't it be possible to simply keep using the `TestingSetup` here, like the tests did previously, and then do specific undo-hacks (hopefully less code) to unload the chainstates (and/or whatever else is required)

Comment (jnewbery) [reply src/test/validation_chainstatemanager_tests.cpp:57]: oops. My mistake. This should be left. I was confusing it with the RPC registration/deregistration.

**Review by promag — COMMENTED** [review]
Concept ACK.

File: src/test/validation_chainstatemanager_tests.cpp
Comment (dongcarl) [reply src/test/validation_chainstatemanager_tests.cpp:51]: @MarcoFalke I assume you mean something like:

```diff
diff --git a/src/test/validation_chainstatemanager_tests.cpp b/src/test/validation_chainstatemanager_tests.cpp
index 34e5723964..664405fc5e 100644
--- a/src/test/validation_chainstatemanager_tests.cpp
+++ b/src/test/validation_chainstatemanager_tests.cpp
@@ -17,48 +17,7 @@
 
 #include <boost/test/unit_test.hpp>
 
-struct ChainTestingSetup : public BasicTestingSetup {
-    boost::thread_group threadGroup;
-
-    explicit ChainTestingSetup()
-        : BasicTestingSetup()
-    {
-        // We have to run a scheduler thread to prevent ActivateBestChain
-        // from blocking due to queue overrun.
-        m_node.scheduler = MakeUnique<CScheduler>();
-        threadGroup.create_thread([&] { TraceThread("scheduler", [&] { m_node.scheduler->serviceQueue(); }); });
-        GetMainSignals().RegisterBackgroundSignalScheduler(*m_node.scheduler);
-
-        pblocktree.reset(new CBlockTreeDB(1 << 20, true));
-
-        m_node.mempool = MakeUnique<CTxMemPool>(&::feeEstimator);
-
-        m_node.chainman = &::g_chainman;
-
-        // Start script-checking threads. Set g_parallel_script_checks to true so they are used.
-        constexpr int script_check_threads = 2;
-        for (int i = 0; i < script_check_threads; ++i) {
-            threadGroup.create_thread([i]() { return ThreadScriptCheck(i); });
-        }
-        g_parallel_script_checks = true;
-    }
-    ~ChainTestingSetup() {
-        if (m_node.scheduler) m_node.scheduler->stop();
-        threadGroup.interrupt_all();
-        threadGroup.join_all();
-        GetMainSignals().FlushBackgroundCallbacks();
-        GetMainSignals().UnregisterBackgroundSignalScheduler();
-        m_node.args = nullptr;
-        UnloadBlockIndex(m_node.mempool.get(), *m_node.chainman);
-        m_node.mempool.reset();
-        m_node.scheduler.reset();
-        m_node.chainman->Reset();
-        m_node.chainman = nullptr;
-        pblocktree.reset();
-    }
-};
-
-BOOST_FIXTURE_TEST_SUITE(validation_chainstatemanager_tests, ChainTestingSetup)
+BOOST_FIXTURE_TEST_SUITE(validation_chainstatemanager_tests, TestingSetup)
 
 //! Basic tests for ChainstateManager.
 //!
@@ -68,6 +27,9 @@ BOOST_AUTO_TEST_CASE(chainstatemanager)
     ChainstateManager& manager = *m_node.chainman;
     CTxMemPool& mempool = *m_node.mempool;
 
+    UnloadBlockIndex(nullptr, manager);
+    manager.Reset();
+
     std::vector<CChainState*> chainstates;
     const CChainParams& chainparams = Params();
 
@@ -151,6 +113,9 @@ BOOST_AUTO_TEST_CASE(chainstatemanager_rebalance_caches)
     ChainstateManager& manager = *m_node.chainman;
     CTxMemPool& mempool = *m_node.mempool;
 
+    UnloadBlockIndex(nullptr, manager);
+    manager.Reset();
+
     size_t max_cache = 10000;
     manager.m_total_coinsdb_cache = max_cache;
     manager.m_total_coinstip_cache = max_cache;
```

I found it hard to convince myself that `UnloadBlockIndex` and `Reset` cleanly resets everything exactly to before `ActivateBestChain` is run, which is why I thought repetition wasn't such a bad thing here. Let me know if I'm missing anything though.

**Comment by dongcarl** [comment]
Addressed a few comments.

-----

fuzz CI failing because of https://github.com/bitcoin/bitcoin/pull/20188#issuecomment-728911018.

**Review by ryanofsky — APPROVED** [review]
Code review ACK 1319b17e15dfd5a3a6bbf3c67b98fcce72c80812. Only changes since last review were suggested simplifications to ChainTestingSetup constructor

**Comment by ryanofsky** [comment]
Hopefully as a test-only cleanup PR with a lot of other changes building on top it (#20158), this can be merged soon

**Comment by MarcoFalke** [comment]
Reason I didn't merge this is that this overlaps with the test changes here: https://github.com/bitcoin/bitcoin/pull/19425#pullrequestreview-534180561. Though there have been outstanding questions and those questions also apply to this pull (in the extent that they overlap).

File: src/test/validation_chainstatemanager_tests.cpp
Comment (ryanofsky) [inline src/test/validation_chainstatemanager_tests.cpp:170]: In commit "test: Add new ChainTestingSetup and use it" (43c880afbcafa9796c838518722594940af69b7d)

Would suggest using `BOOST_REQUIRE` instead of `assert` or splitting as `bool loaded = c1.LoadGenesisBlock(Params()); assert(loaded);`. Writing code with side effects in an `assert` adds some ambiguity because behavior will depend on value of NDEBUG

**Review by ryanofsky — COMMENTED** [review]
> Reason I didn't merge this is that this overlaps with the test changes here: [#19425 (review)](https://github.com/bitcoin/bitcoin/pull/19425#pullrequestreview-534180561). Though there have been outstanding questions and those questions also apply to this pull (in the extent that they overlap).

Thanks. I guess I need to rebase #19425. I'll check out the test changes too there and make updates if needed to make sure they are compatible with this PR.

This PR just takes approach of having fewer duplicate disconnected state objects inside tests, so I don't think there are real questions about it or major drawbacks. But if anyone has concerns or questions (even vague ones), it'd be good to post something so Carl can reply and this work doesn't languish

File: src/test/validation_chainstatemanager_tests.cpp
Comment (dongcarl) [reply src/test/validation_chainstatemanager_tests.cpp:170]: Oh! Didn't know about `NDEBUG`... So when `NDEBUG` is set, the code inside `assert(...)` may be optimized out? Or is it the case that only the result isn't checked/asserted?

**Comment by jnewbery** [comment]
utACK 1319b17e15dfd5a3a6bbf3c67b98fcce72c80812

**Comment by dongcarl** [comment]
Pushed 1319b17e15dfd5a3a6bbf3c67b98fcce72c80812 -> 11fac8b2c8001e9aee9debb20df09473b9092b8c:
1. Rebased over master for #20222
2. Changed `assert`s in tests to `BOOST_REQUIRE` as recommended: https://github.com/bitcoin/bitcoin/pull/20323#discussion_r533584267
3. Extracted out testing app initialization and destruction logic into `TestingAppInitSequence` and `TestingAppDestructionSequence` to improve sync and decrease maintenance burden between `ChainTestingSetup` and `TestingSetup` to address: https://github.com/bitcoin/bitcoin/pull/20323#discussion_r520810288

In (3), ordering-wise, the script-checking thread initialization was moved up to before the chainstate gets initialized and RPC command registration was moved down. Neither of which should have any tangible effect on the correctness of initialization.

**Comment by jnewbery** [comment]
I've had a quick look through the new branch and find the `TestingAppInitSequence()` and `TestingAppDestructionSequence()` less clear than the previous version. In general functions that take a bool argument which substantially changes the logic inside that function (in this case `chain_testing_only`), suggest that the abstraction may be wrong.

Was there an actual problem with the previous branch, or are you just trying to reduce code duplication?

File: src/test/util/setup_common.cpp
Comment (MarcoFalke) [inline src/test/util/setup_common.cpp:188]: any reason for the modifications here? reset is idempotent , so doesn't need to be guarded.

File: src/test/validation_chainstatemanager_tests.cpp
Comment (MarcoFalke) [inline src/test/validation_chainstatemanager_tests.cpp:20]: Any reason to add this here and not to setup_common? Moving it to setup_common would allow removing the confusing `TestingAppInitSequence` declarations in the header. Also, it would allow other tests to use the `ChainTestingSetup`.

File: src/test/util/setup_common.cpp
Comment (MarcoFalke) [inline src/test/util/setup_common.cpp:147]: Agree with john that functions that have a large branch (and a boolean arg) are better split off into two functions

**Review by maflcko — COMMENTED** [review]
left some comments

**Review by ryanofsky — APPROVED** [review]
Code review ACK 11fac8b2c8001e9aee9debb20df09473b9092b8c. Changes since last review are deduping test init functions and replacing asserts (thanks!).

Other review suggestions for splitting up functions instead of using a bool arguments seem reasonable. 

Another idea (not sure whether it is good or applicable): an alternative to calling pairs of functions in constructors and destructors is to just add class members. Member constructors will run in containing class's default constructor, and same with destructors. Seems like this would allow getting rid of the 4 one-line constructor/destructors.

**Comment by dongcarl** [comment]
Thanks everyone for the reviews. The latest push 81137c6 is now rebased over #19485.

This PR is intended to just move things around a bit for the rest of the `ChainstateManager` de-globalization to work properly. I believe we've gotten a bit stuck on the intricacies of these `TestingSetup` classes.

Personally, I don't particularly care how the abstraction works here. I see the `TestingSetup` classes and their inheritance hierarchy as a poor abstraction of how our initialization and de-initialization logics (barely) work. Therefore, I'd like to just fix this particular problem for now and move on.

I do fully intend on coming back to the initialization codepaths later, most likely after having moved a couple of mutable statics into classes they logically belong in, and having a larger PR that coalesces and cleans all of this up.

**Review by ryanofsky — APPROVED** [review]
Code review ACK 81137c60fe6234569e1c5e6760f3a6f016956944. This change is simpler after the rebase because wallet & bench commits are dropped.

Along with the rebase, new test init functions were dropped again in response to https://github.com/bitcoin/bitcoin/pull/20323#issuecomment-739830876, which is fine. I think in the longer run more init code can be moved into RAII classes, and we won't need these intricate initialization functions and class hierarchies (tests could just state their resource requirements as variable declarations).

**Comment by MarcoFalke** [comment]
Thanks for shuffling this around for over a month

ACK 81137c60fe looking excellent now 🐩

<details><summary>Show signature and timestamp</summary>

Signature:

```
-----BEGIN PGP SIGNED MESSAGE-----
Hash: SHA512

ACK 81137c60fe looking excellent now 🐩
-----BEGIN PGP SIGNATURE-----

iQGzBAEBCgAdFiEE+rVPoUahrI9sLGYTzit1aX5ppUgFAlwqrYAACgkQzit1aX5p
pUjLNQv9G0SWcw2rzItf1EIttkPPy6CX5BxM3WRp4MY+hW42LikQIcTm236LyD9t
/e4homRxM0DT1q6/JKin3RP4y1O4zpLswFaz8nNH7ScmLGyTNdb/HUDxLLNqnynb
TYZYpZP6Md/EWjCoF2kKEBlVmzUeehagq3L9gQ7udXTMZtMu2z1jP6v7/TLe6XxA
HqABCQn9BJYcD5dqHETg4TYBP2hG5bUnh5rmJWHnn0ifdEmnS4cbExtHCL9nxQpF
1evmfgdk6nXzkkEllJsOm2tYVGhacW5Z483w/jFveBPmVJgduYmAO5mddMreqV/5
/1Gegw1YQnb6LmEyHhsD2POxlidXksNTOEAyghOwylFFPN8467R/LQEXr8ltstQ9
NnhkoCNVgSDBuzkdTB+n8S6gfQyTp2T3klrqr3GHJIPlRikRLGI5bOwtjtZsFkGX
j5508lSIANRoUKGv3nbfWL47l8he+dYMfCvpqkLWxRPqGT+MnIQV9aRKKxLcSFIb
czC392Ra
=hAKu
-----END PGP SIGNATURE-----
```

[Timestamp of file with hash `1fd492318f7f8e7f0f9f62fce5042d824a2c5c3fcb368e2027f02582ea8ed27e  -`](https://opentimestamps.org/info/?004f70656e54696d657374616d7073000050726f6f6600bf89e2e884e8929401081fd492318f7f8e7f0f9f62fce5042d824a2c5c3fcb368e2027f02582ea8ed27ef0103e3593fed7da326e22d4c7cab47b6a5808fff01060f86268a2ab50023831207e0d17d5c408f1045fd07539f0081526a420059384e10083dfe30d2ef90c8e232268747470733a2f2f6274632e63616c656e6461722e636174616c6c6178792e636f6dfff0108a21a35dba35cb523c2d87be010dd57c08f1045fd07539f008e8d8c81035886ace0083dfe30d2ef90c8e292868747470733a2f2f66696e6e65792e63616c656e6461722e657465726e69747977616c6c2e636f6dfff010a7cb02798d158bbec4791e51dc25387f08f1045fd07539f0089607b427ecf873830083dfe30d2ef90c8e2c2b68747470733a2f2f626f622e6274632e63616c656e6461722e6f70656e74696d657374616d70732e6f7267f010cf5aa2e6831f77a75373d16afe36cd4f08f1045fd07539f008f6d3a6b761ae1a890083dfe30d2ef90c8e2e2d68747470733a2f2f616c6963652e6274632e63616c656e6461722e6f70656e74696d657374616d70732e6f7267)

</details>

**Comment by MarcoFalke** [comment]
> 2. Put ChainTestingSetup back in setup_common so that future tests can use it as requested here. (Sorry John, I knew you weren't the biggest fan of this confused)

Ok, let's wait for @jnewbery 's green light before merge

**Comment by jnewbery** [comment]
ACK 81137c60fe6234569e1c5e6760f3a6f016956944

**Pull request merged**
